### PR TITLE
`Discussion`: Reference artifacts within postings

### PR DESCRIPTION
**Pull request by canberkanar**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
In the current context, users refer to artifacts of Artemis such as lectures, attachments, and exercises byword in their postings. Due to this, others who want to get an idea of the mentioned artifact within a discussion first have to manually find and navigate to the artifact to access it. 

### Description
To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'. When clicked, the list of the Exercises and Lectures of the current course is displayed. Attachments are also displayed in a nested way, where each is located under their specific Lecture. When a user selects an artifact, the corresponding markdown code is automatically generated and inserted into the markdown editor. Later when in preview mode or the posting is just viewed, the attachment's name is displayed as a link, together with an icon with respect to the type of the attachment.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Course with postings enabled

1. Log in to Artemis
2. Navigate to Course Discussions page
3. Create/edit postings with references to types (check the screencast for help)
- Post (can reference with : #{POST_ID} )
- File-upload Exercise
- Programming Exercise
- Modeling Exercise
- Text Exercise
- Quiz Exercise
- Lecture
- Lecture Attachment
4.  Verify that unique icons are displayed for each type of artifact referenced and make sure that they are consistent with the icons used in Artemis for those types.
5. Verify that previewing within the markdown editor works as well.
6. Make sure that all types of references are accessed when clicked.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
  exerciseReferenceCommand.ts | 100% | 100% |                                                                
  lectureAttachmentReferenceCommand.ts | 90% | 100% |
  metis.service.ts | 78.7% |  92.54% |
  metis.util.ts | 100% | 100% |                                                          
  posting-content.components.ts | 82.45% | 97.95% |                                    
  posting-content-part.components.ts | 91.66% | 100% | 
  posting-markdown-editor.component.ts | 77.77% | 92.85% |                                    
  negated-type-check.pipe.ts | 100% | 100% |                                                          

### Screenshots (UPDATED!!!)

https://user-images.githubusercontent.com/67344158/174290816-0e3c5777-36ab-417e-aac6-b94ff07af3bb.mp4

### CODE REVIEW

**Review by jbllhff — DISMISSED** [review]
Tested on ts2, works great, thanks!

**Review by akesfeden — CHANGES_REQUESTED** [review]
Tested on TS1, references have correct icons and navigation works as expected, however Preview tab does not remove the closing tags when displaying them. I added screenshots below.

image

image

**Review by TheTrueVester — DISMISSED** [review]
Tested on ts1. Everything worked as expected.

File: src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts [hunk 1]
```diff
@@ -56,14 +56,33 @@ export class PostingContentComponent implements OnInit, OnDestroy {
...
+                    queryParams = referencedPostInLoadedPosts ? this.metisService.getQueryParamsForPost(referencedPostInLoadedPosts) : ({ searchText: referenceStr } as Params);
+                } else if (
+                    ReferenceType.LECTURE === referenceType ||
+                    ReferenceType.PROGRAMMING === referenceType ||
+                    ReferenceType.MODELING === referenceType ||
+                    ReferenceType.QUIZ === referenceType ||
+                    ReferenceType.TEXT === referenceType ||
+                    ReferenceType.FILE_UPLOAD === referenceType
+                ) {
+                    referenceStr = this.content!.substring(this.content?.indexOf(')', patternMatch.startIndex)! + 1, patternMatch.endIndex - (referenceType.length + 3));
```
Comment (Dominik-Weinzierl) [inline src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:80]: Could you please add some comments and an example to understand the string magic? :D e.g. why do you add 3, ...

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 2]
```diff
@@ -1,30 +1,60 @@
...
+                { provide: Router, useClass: MockRouter },
+            ],
         })
             .compileComponents()
             .then(() => {
                 fixture = TestBed.createComponent(PostingContentPartComponent);
                 component = fixture.componentInstance;
                 debugElement = fixture.debugElement;
+                router = TestBed.inject(Router);
+                metisService = TestBed.inject(MetisService);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:50]: image

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 3]
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
...
             expect(referenceLink.innerHTML).toInclude(referenceStr);
+
+            // should display relevant icon for post
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-message');
+
+            // on click should navigate to referenced post within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:113]:
```suggestion
            expect(navigateByUrlSpy).toHaveBeenCalledOnce();
```

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 4]
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
...
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:114]:
```suggestion
            expect(openAttachmentSpy).not.toHaveBeenCalled();
```

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 5]
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
...
+
+            // should display relevant icon for file upload exercise
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-file-arrow-up');
+
+            // on click should navigate to referenced file upload exercise within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:141]: again

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 6]
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
...
+
+            // should display relevant icon for modeling exercise
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-diagram-project');
+
+            // on click should navigate to referenced modeling exercise within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:168]: again

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 7]
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
...
+
+            // should display relevant icon for quiz exercise
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-check-double');
+
+            // on click should navigate to referenced quiz exercise within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:195]: again

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts [hunk 8]
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
...
+
+            // should display relevant icon for attachment
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-file');
+
+            // on click should open referenced attachment within new tab
+            referenceLink.click();
+            expect(openAttachmentSpy).toBeCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledWith(attachmentURL);
```
Comment (Dominik-Weinzierl) [inline src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:331]: Please replace those tests with parameterized tests (this should be done really easy)

**Review by Dominik-Weinzierl — CHANGES_REQUESTED** [review]
Code looks good - some minor suggestions

File: src/main/webapp/app/shared/metis/metis.service.ts [hunk 9]
```diff
@@ -356,6 +356,34 @@ export class MetisService implements OnDestroy {
...
+    }
+
+    /**
+     * returns the router link required for navigating to the lecture referenced within a posting
+     * @param {string} lectureId ID of the lecture to be navigated to
+     * @return {string} router link of the lecture
+     */
+    getLinkForLecture(lectureId: string): string {
+        return '/courses/' + this.getCourse().id + '/lectures/' + lectureId;
+    }
```
[the remark this answers is not in the document]
Comment (canberkanar) [reply src/main/webapp/app/shared/metis/metis.service.ts:375]: Good catch! I missed this one while refactoring code it seems.

File: src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts (hunk 8 above)
Comment (canberkanar) [reply src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:331]: Useful concept, thanks!

**Review by TheTrueVester — DISMISSED** [review]
Retested on ts1. Everything worked as expected.

**Review by Dominik-Weinzierl — APPROVED** [review]
Code changes looks good to me :+1: thank you for implementing my suggestions and for adapting the tests :+1:

**Review by alaraozdenler — APPROVED** [review]
Tested on ts2, everything works as described

**Comment by ge65cer** [comment]
Great work! 👍 
Some suggestions: Would it be possible to also display the exercise icons in the dropdown list? Also, I'd be awesome if the dropdown is scrollable for courses with many exercises and lectures (currently, the whole modal scrolls up).

**Comment by canberkanar** [comment]
> Great work! 👍 Some suggestions: Would it be possible to also display the exercise icons in the dropdown list? Also, I'd be awesome if the dropdown is scrollable for courses with many exercises and lectures (currently, the whole modal scrolls up).

Thank you for your review. Since the drop-down part is in the scope of markdown-editor and is designed in reusable way, the feature of supporting scrolling would be out of the scope of this PR. For instance, task and test cases dropdowns seen while editing the description of programming exercises are also allowed while using this feature. Adding support for icons is also not feasible in this shared component,.

**Review by minrows — DISMISSED** [review]
Tested on ts2. LGTM 👍

**Review by akesfeden — DISMISSED** [review]
Previous issues are solved, thank you for the fix! Working as expected and looks good.

**Review by ge65cer — APPROVED** [review]

**Review by Hialus — CHANGES_REQUESTED** [review]
Tested on TS1 and I found some usability issues:
When creating a post for an exercise the model pops up and you have to scroll down the entire modal to select an exercise:
grafik

grafik

This is even worse when it happens outside of a modal:
grafik

You should consider making the dropdown content itself scrollable here.

Additionally, I don't like the usage of a simple dropdown for this. Especially if you have a lot of exercises (as seen in the screenshots above), it is very hard to find the exercise you are looking for. A search bar would help a lot here.

Another issue with this dropdown is its displaying on mobile devices. The dropdown can be cutoff as seen here:
grafik
This could be fixed by dynamically aligning the content as explained here: [https://getbootstrap.com/docs/5.0/components/dropdowns/#responsive-alignment](https://getbootstrap.com/docs/5.0/components/dropdowns/#responsive-alignment) 

Minor issue: The `[programming]` tag is blue while the `[/programming]` is not 
grafik

While the last issue can be ignored, if it is not easy to fix, the first three cannot. They make this feature very user-unfriendly and should be fixed in this PR and not a follow-up.

**Review by Dominik-Weinzierl — CHANGES_REQUESTED** [review]
I agree with @Hialus - you can provide a simple scrollable dropdown and add a search bar in a follow up for example (if this is too much for this PR)

**Comment by canberkanar** [comment]
> Another issue with this dropdown is its displaying on mobile devices. The dropdown can be cutoff as seen here:

Thanks for your feedback.

- I added scrolling to the drop-down. I wasn't aware of the problem with mobile devices, yet it seems to me that the problem is solved there as well.
- The patterns I used to encode the references overlapped with Markup's own URL referencing command. That was the reason for the first section being in a different color than the trailer. Although I don't think this was really an issue, I changed the patterns so that some users are not distracted as in this case.
- It is up to you to approve this PR or not, but I am not adding a search function here. The exercise referencing feature was already merged and this "user-unfriendly" feature is already present in the code base. I will discuss this with Stephan when the time comes and will do the necessities in a follow-up PR later.

**Comment by canberkanar** [comment]
> I agree with @Hialus - you can provide a simple scrollable dropdown and add a search bar in a follow up for example (if this is too much for this PR)

Thanks! I added scrolling to the dropdown menu.

**Comment by Dominik-Weinzierl** [comment]
Thanks! Works as expected :+1: some suggestions and additional issues
- I think it's enough if you either highlight or underline the selected element. :thinking: 
image
- The lists (and quotes) do not work as expected
 image
 image

**Review by nadinesamyy — DISMISSED** [review]
Works as expected!

**Comment by Hialus** [comment]
> The exercise referencing feature was already merged and this "user-unfriendly" feature is already present in the code base. 

Your PR description says otherwise, which is why I remarked on it.
> To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'. 

But the same issue would also apply to the lecture dropdown, as e.g. EIST also has a lot of lecture attachments, which would, again, make it hard for a user to find the one they want. This is by definition user-unfriendly. If I were to write a post referencing an exercise or lecture I would continue to just write the name of it instead of using this feature, as it would be quite a bit faster and simpler.
There are a couple of ways to make this feature more user-friendly:
- Sort the list. Currently, it appears to be either unsorted or sorted by ID.
- Provide a search bar in the dropdown
- Provide a way to add references while you are typing (with an auto-completion feature like used for test case names in Artemis or for PR/issue references on GitHub)
- Replace the dropdown with something else. E.g:
    - A modal
    - A sidebar (maybe with drag and drop functionality)
- Automatically add a reference when a user posts a link to an exercise/lecture/attachment

I still think that you should address this already in this PR and only create a follow-up PR to further improve it. While it is unfortunate that the exercise referencing feature has already been merged the PR description currently describes it to at least be modified by this PR. Therefore changing this would not be out of scope currently.

**Comment by canberkanar** [comment]
> > The exercise referencing feature was already merged and this "user-unfriendly" feature is already present in the code base.
> 
> Your PR description says otherwise, which is why I remarked on it.
> 
> > To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'.
> 
> But the same issue would also apply to the lecture dropdown, as e.g. EIST also has a lot of lecture attachments, which would, again, make it hard for a user to find the one they want. This is by definition user-unfriendly. If I were to write a post referencing an exercise or lecture I would continue to just write the name of it instead of using this feature, as it would be quite a bit faster and simpler. There are a couple of ways to make this feature more user-friendly:
> 
> * Sort the list. Currently, it appears to be either unsorted or sorted by ID.
> * Provide a search bar in the dropdown
> * Provide a way to add references while you are typing (with an auto-completion feature like used for test case names in Artemis or for PR/issue references on GitHub)
> * Replace the dropdown with something else. E.g:
>   
>   * A modal
>   * A sidebar (maybe with drag and drop functionality)
> * Automatically add a reference when a user posts a link to an exercise/lecture/attachment
> 
> I still think that you should address this already in this PR and only create a follow-up PR to further improve it. While it is unfortunate that the exercise referencing feature has already been merged the PR description currently describes it to at least be modified by this PR. Therefore changing this would not be out of scope currently.

Yes as you have mentioned, the exercise referencing is modified in this PR. When developing, I try to be consistent with what has been already merged in Artemis. Take a look at the screenshot below. I wonder if the drop-down here is not user-friendly, how did a very similar drop-down exists in production. This is very that reviewers' opinions are relative.

Untitled

**Comment by Hialus** [comment]
> This is very sad that reviewers' opinions are relative.

That's just human nature. Reviews are always subjective to some degree, and reviewers look for different things when testing a feature (E.g., some reviewers only test if the testing steps work as described). This is also why more than one manual test is required to merge a PR.

And to answer your question: Yes, I do think that dropdown is user-unfriendly as well. It would be faster for me to directly go to the exercise page, than to go to this page and then select the exercise I want.

**Comment by canberkanar** [comment]
> > * Provide a way to add references while you are typing (with an auto-completion feature like used for test case names in Artemis or for PR/issue references on GitHub)

I will check if this is feasible. For test cases, it works very well, but we have to check if it would be performance-wise okay searching through all the lectures, exercises, and attachments with each keyboard click.

**Review by TheTrueVester — APPROVED** [review]
Retested on ts1. Everything worked as expected.

**Review by max-bergmann — APPROVED** [review]
Tested on TS1, works as expected

**Review by SimonPaul02 — APPROVED** [review]
Tested on TS 2, works as expected. But I guess the students won't know how to reference other posts. It would be nice have some kind of dropdown option for posts. But this is not part of this PR, good job!

**Review by yuling-sun — APPROVED** [review]
Tested on TS2. References work as expected. It's not intuitive that references to other posts are done with #{POST_ID}.

**Comment by canberkanar** [comment]
> Tested on TS 2, works as expected. But I guess the students won't know how to reference other posts. It would be nice have some kind of dropdown option for posts. But this is not part of this PR, good job!

Thanks! Referencing posts was part of a previous development, but we will consider your opinion for future improvements. 👍🏻

**Comment by ole-ve** [comment]
Nice feature.
Regarding the lecture references, I'd have two comments:
Bildschirmfoto 2022-06-21 um 11 10 53

1. To me as a user, it is not clear which entry of `Lecture 2` to choose. Did you simply create multiple lectures with the same name or multiple lecture units with the same name? In addition, a small icon for every entry that displays if the entry represents a lecture/attachment may also help.
2. Agreeing to Timor, a dropdown can become user-unfriendly very fast. Especially, when there are a lot of lectures/attachments (in EIST: 14 Lectures á 5 Slidesets = 70 Attachments). However, using a modal makes the process of adding references more complex and may not be used as a result. Therefore, I'd suggest adding a search bar to query for units.

**Review by babyygemperor — APPROVED** [review]
Tested on ts2, works as described

**Comment by canberkanar** [comment]
> Thanks! Works as expected 👍 some suggestions and additional issues
> 
> * I think it's enough if you either highlight or underline the selected element. 🤔
>   image
> * The lists (and quotes) do not work as expected
>   image
>   image

Thanks for the feedback once again.

1.  I have removed the underlining on hover.
2. The problem is not in scope of this PR as the same problem can be replicated by referencing a post within an ordered list as well. Here is the issue I have created for your reference, which will be fixed in the future. #5281 We discussed the issue with Stephan today and decided that this PR can be merged.

**Comment by canberkanar** [comment]
> Nice feature. Regarding the lecture references, I'd have two comments: Bildschirmfoto 2022-06-21 um 11 10 53
> 
> 1. To me as a user, it is not clear which entry of `Lecture 2` to choose. Did you simply create multiple lectures with the same name or multiple lecture units with the same name? In addition, a small icon for every entry that displays if the entry represents a lecture/attachment may also help.
> 2. Agreeing to Timor, a dropdown can become user-unfriendly very fast. Especially, when there are a lot of lectures/attachments (in EIST: 14 Lectures á 5 Slidesets = 70 Attachments). However, using a modal makes the process of adding references more complex and may not be used as a result. Therefore, I'd suggest adding a search bar to query for units.

1. Multiple lectures with the same name were created for testing purposes.
2. We will consider your suggestion for future improvements. Thanks!

**Pull request merged**